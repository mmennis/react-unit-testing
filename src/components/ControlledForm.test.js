import React from 'react';
import { shallow } from 'enzyme';
import ControlledForm from './ControlledForm';

describe('ControlledForm', () => {
    let wrapper;
    let mockSubmit;

    beforeEach(() => {
        mockSubmit = jest.fn();
        wrapper = shallow(<ControlledForm submit={mockSubmit} />)
    });

    it('should match he snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })

    describe('handleChamge', () => {
        it('should call setState on title', () => {
            const mockEvent = {
                target: {
                    name: "title",
                    value: "test"
                }
            };

            const expected = {
                title: "test",
                description: "",
                submitActive: false
            };

            wrapper.instance().handleChange(mockEvent);
            expect(wrapper.state()).toEqual(expected);
        })

        it('should call setState on description', () => {
            const mockEvent = {
                target: {
                    name: "description",
                    value: "test"
                }
            };

            const expected = {
                title: "",
                description: "test",
                submitActive: false
            };

            wrapper.instance().handleChange(mockEvent);
            expect(wrapper.state()).toEqual(expected);
        })

        it('should call checkFields', () => {
            const spy = jest.spyOn(wrapper.instance(), "checkFields");
            wrapper.instance().forceUpdate();

            const mockEvent = {
                target: {
                    name: "description",
                    value: "test"
                }
            }
            const expected = true;
            wrapper.instance().handleChange(mockEvent);
            expect(spy).toHaveBeenCalled();
        })

        it('should call handleChange on title change with the correct params', () => {
            const spy = jest.spyOn(wrapper.instance(), "handleChange");
            wrapper.instance().forceUpdate();

            const mockEvent = {
                target: {
                    name: "title",
                    value: "test"
                }
            }

            wrapper.find('.title-input').simulate("change", mockEvent);
            expect(spy).toHaveBeenCalledWith(mockEvent);
        })

        it('should call handleChange on description change with the correct params', () => {
            const spy = jest.spyOn(wrapper.instance(), "handleChange");
            wrapper.instance().forceUpdate();

            const mockEvent = {
                target: {
                    name: "description",
                    value: "test"
                }
            };

            wrapper.find(".description-input").simulate("change", mockEvent);
            expect(spy).toHaveBeenCalledWith(mockEvent);
        })
    })
    
    describe('handleSubmit', () => {
        it('should call preventDefault', () => {
            const mockPreventDefault = jest.fn();

            const mockEvent = {
                preventDefault: mockPreventDefault
            }
            wrapper.instance().handleSubmit(mockEvent);
            expect(mockPreventDefault).toHaveBeenCalled();
        })

        it('should return if submitActive is false', () => {
            const mockPreventDefault = jest.fn();
            const mockEvent = {
                preventDefault: mockPreventDefault
            }

            const spy = jest.spyOn(wrapper.instance(), "handleSubmit");
            wrapper.instance().forceUpdate();

            wrapper.instance().handleSubmit(mockEvent);
            expect(spy).toReturn();
        })

        it('should call submit with the right params', () => {
            wrapper.setState({
                title: "test title",
                description: "test description",
                submitActive: true
            });

            const expected = {
                title: "test title",
                description: "test description"
            };

            const mockPreventDefault = jest.fn();

            const mockEvent = {
                preventDefault: mockPreventDefault
            };
            wrapper.instance().handleSubmit(mockEvent);
            expect(mockSubmit).toHaveBeenCalledWith(expected);
        })

        // FIXME - no idea why this is broken
        it.skip('should call handleSubmit on submit with the correct params', () => {
            const spy = jest.spyOn(wrapper.instance(), "handleSubmit");
            wrapper.instance().forceUpdate();

            const mockPreventDefault = jest.fn();

            const mockEvent = {
                preventDefault: mockPreventDefault
            };

            wrapper.find('.controlled-form').simulate("submit", mockEvent);
            expect(spy).toHaveBeenCalled()
        })
    })

    describe('checkFields', () => {
        it('should setState if it meets the conditions', () => {
            wrapper.setState({
                title: "test title",
                description: "test description",
                submitActive: false
            });
            wrapper.instance().checkFields();

            expect(wrapper.state().submitActive).toEqual(true);
        })

        it('should setState if it does not meet the conditions', () => {
            wrapper.setState({
                title: "test title",
                description: "",
                submitActive: true
            })

            wrapper.instance().checkFields();

            expect(wrapper.state().submitActive).toEqual(false);
        })
    })
});