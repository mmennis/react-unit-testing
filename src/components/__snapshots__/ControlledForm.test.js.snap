// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlledForm should match he snapshot 1`] = `ShallowWrapper {}`;
